src/widgets/scrollbar.rs

```diff
@@ -0,0 +1,853 @@
+    fn is_vertical(&self) -> bool {
+        match self.orientation {
+            ScrollbarOrientation::VerticalRight | ScrollbarOrientation::VerticalLeft => true,
+            ScrollbarOrientation::HorizontalBottom | ScrollbarOrientation::HorizontalTop => false,
+        }
+    }
+
+    fn get_track_area(&self, area: Rect) -> Rect {
+        // Decrease track area if a begin arrow is present
+        let area = if self.begin_symbol.is_some() {
+            if self.is_vertical() {
+                // For vertical scrollbar, reduce the height by one
+                Rect::new(
+                    area.x,
+                    area.y + 1,
+                    area.width,
+                    area.height.saturating_sub(1),
+                )
+            } else {
+                // For horizontal scrollbar, reduce the width by one
+                Rect::new(
+                    area.x + 1,
+                    area.y,
+                    area.width.saturating_sub(1),
+                    area.height,
+                )
+            }
+        } else {
+            area
+        };
+        // Further decrease scrollbar area if an end arrow is present
+        if self.end_symbol.is_some() {
+            if self.is_vertical() {
+                // For vertical scrollbar, reduce the height by one
+                Rect::new(area.x, area.y, area.width, area.height.saturating_sub(1))
+            } else {
+                // For horizontal scrollbar, reduce the width by one
+                Rect::new(area.x, area.y, area.width.saturating_sub(1), area.height)
+            }
+        } else {
+            area
+        }
+    }
+
+    fn should_not_render(&self, track_start: u16, track_end: u16, content_length: u16) -> bool {
+        if track_end - track_start == 0 || content_length == 0 {
+            return true;
+        }
+        false
+    }
+
+    fn get_track_start_end(&self, area: Rect) -> (u16, u16, u16) {
+        match self.orientation {
+            ScrollbarOrientation::VerticalRight => {
+                (area.top(), area.bottom(), area.right().saturating_sub(1))
+            }
+            ScrollbarOrientation::VerticalLeft => (area.top(), area.bottom(), area.left()),
+            ScrollbarOrientation::HorizontalBottom => {
+                (area.left(), area.right(), area.bottom().saturating_sub(1))
+            }
+            ScrollbarOrientation::HorizontalTop => (area.left(), area.right(), area.top()),
+        }
+    }
+
+    fn get_thumb_start_end(
+        &self,
+        state_content_length: u16,
+        state_position: u16,
+        viewport_content_length: u16,
+        track_start: u16,
+        track_end: u16,
+    ) -> (u16, u16) {
+        let scroll_position_ratio = (state_position as f64 / state_content_length as f64).min(1.0);
+
+        let thumb_size = (((viewport_content_length as f64 / state_content_length as f64)
+            * (track_end - track_start) as f64)
+            .round() as u16)
+            .max(1);
```

Comment:
```suggestion
        fn get_thumb_start_end(
        &self,
        state: &ScrollbarState,
        viewport_content_length: u16,
        track_start: u16,
        track_end: u16,
    ) -> (u16, u16) {
        let scroll_position_ratio = (state.position as f64 / state.content_length as f64).min(1.0);

        let thumb_size = (((viewport_content_length as f64 / state.content_length as f64)
            * (track_end - track_start) as f64)
            .round() as u16)
            .max(1);
```

Againt nit pick, but I'd simplify the arg list a tad bit

Reply:
Resolved in https://github.com/tui-rs-revival/ratatui/pull/228/commits/2284b202bd71c2e1c0549aaa37a89276aeee2c42.